datavec/datavec-python/src/main/java/org/datavec/python/PythonExecutioner.java

```diff
@@ -0,0 +1,613 @@
+package org.datavec.python;
+
+import java.io.File;
+import java.io.FileInputStream;
+import java.lang.reflect.Array;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.regex.Pattern;
+
+import lombok.extern.slf4j.Slf4j;
+import org.json.simple.JSONArray;
+import org.json.simple.parser.JSONParser;
+import org.bytedeco.javacpp.*;
+import static org.bytedeco.javacpp.python.*;
+
+import org.nd4j.linalg.api.buffer.DataType;
+
+@Slf4j
+public class PythonExecutioner {
+    private String name;
+    private Pointer namePtr;
+    private boolean restricted = false;
+    private PyObject module;
+    private PyObject globals;
+    private JSONParser parser = new JSONParser();
+    private Map<String, PyThreadState> interpreters = new HashMap<String, PyThreadState>();
+    private String currentInterpreter =  null;
+    private static PythonExecutioner pythonExecutioner;
+    private boolean safeExecFlag = false;
+
+    public static PythonExecutioner getInstance(){
+        // do not use constructor
+        if (pythonExecutioner == null){
+            pythonExecutioner = new PythonExecutioner();
+        }
+        return pythonExecutioner;
+    }
+    public void setInterpreter(String name){
+        if (name == null){
+            if (currentInterpreter != null){
+                PyThreadState_Swap(null);
+                currentInterpreter = null;
+            }
+        }
+        if (currentInterpreter != null && currentInterpreter.equals(name)){
+            return;
+        }
+        else if (interpreters.containsKey(name)){
+            PyThreadState threadState = interpreters.get(name);
+            log.info("CPython: PyThreadState_Swap()");
+            PyThreadState_Swap(threadState);
+            init();
+        }
+        else{
+            log.info("CPython: Py_NewInterpreter()");
+            PyThreadState threadState = Py_NewInterpreter();
+            interpreters.put(name, threadState);
+            log.info("CPython: PyThreadState_Swap()");
+            PyThreadState_Swap(threadState);
+            init();
+        }
+        currentInterpreter = name;
```

Comment:
Does using multiple Python interpreters actually help? IIRC, the GIL applies to the process as a whole, not for each interpreter, or did that change?

Reply:
But we can have multiple "contexts" in memory. There are tests for this. (See TestSamdbox.java)